### ISSUE TITLE
Routing celery logs through structlog

### ISSUE THREAD

**Issue by Charlie-iProov** [hynek/structlog#287]
```py
class Logger:
    def __new__(cls, *args, **kwargs):
        logging.basicConfig(level=os.getenv("LOG_LEVEL", "DEBUG").upper(), format="%(message)s")
            structlog.configure(
                processors=[
                    add_module_and_lineno,
                    structlog.stdlib.filter_by_level,
                    structlog.stdlib.add_log_level,
                    structlog.stdlib.PositionalArgumentsFormatter(),
                    structlog.processors.TimeStamper(fmt="iso", key="ts"),
                    structlog.processors.format_exc_info,
                    structlog.processors.UnicodeDecoder(),
                    structlog.processors.JSONRenderer(sort_keys=True, serializer=rapidjson.dumps),
                    structlog.stdlib.ProcessorFormatter.wrap_for_formatter,
                ],
                context_class=dict,
                logger_factory=structlog.stdlib.LoggerFactory(),
                wrapper_class=structlog.stdlib.BoundLogger,
                cache_logger_on_first_use=True,
            )
            return get_logger()
```

I have a logger class which I use very frequently. This generates really nice json logs most of the time. However, I'm struggling to get this working with celery. It seems celery has its own logging module which I cant seem to route through structlog. How would one route the logging output of a library which has its own logger through the class that im currently using? I've read through your documentation but I seem to be lacking some fundamental understanding here.

Thanks in advance.

**Comment by hynek** [comment]
Since celery is using stdlib `logging`, I think the cleanest way is to pipe stdlib into structlog: https://www.structlog.org/en/stable/standard-library.html#rendering-using-structlog-based-formatters-within-logging

**Comment by f3ndot** [comment]
The challenge with this advice is that, when following the Celery documentation of using `logger = get_task_logger(__name__)` you get not a structlog logger, but rather the regular stdlib `logging.Logger`. Your tasks cannot benefit from all of the features of structlog, and you must either suffer regular logging or use `extra={}` to pass structured fields in.

Advice is slim on whether or not its acceptable to not use `get_task_logger(__name__)` and instead prefer `structlog.get_logger(__name__)` inside your tasks. It seems to suggest there is thread safety or other such Celery internal conventions that may become unstable if you don't use the task logger (please correct me if I'm wrong).

Peeking at [`django-structlog`](https://github.com/jrobichaud/django-structlog)'s source code to determine their approach, they appear to actually use `structlog.getLogger()` inside their task functions but they also heavily modify the state of the world with Celery signals. I am unsure if this is the most "blessed" approach or needlessly complex.

At the end of the day it would be great to be able to do:

```python3
@celery.task()
def my_task(some_arg):
    logger.info("Some log entry", some_arg=some_arg, foo="bar")
```

And get the same formatting as if I did a `logger.info` in the rest of my code.

**Comment by hynek** [comment]
I suppose if celery insists on using `get_task_logger()` you can either write a LoggerFactory like https://github.com/hynek/structlog/blob/3e28d6530fc96f1d8691e442644c18b6f3a1463a/src/structlog/stdlib.py#L514-L556

 and just replace `logging.getLogger` with `get_task_logger`.

Or you handle it in your own code and use `structlog.wrap_logger(get_task_logger())`

**Comment by f3ndot** [comment]
@hynek I really appreciate you replying (and so quickly too!) and your support. Taking more time to learn about the Python stdlib `logging` fundamentals I understand more clearly how naming loggers work along with propagation. For various reasons, I already have stdlib piping into structlog, and use a dictConfig to accomplish that. The missing bit was twofold:

1. Defining some logging configuration for the special `celery.task` logger in my dictConfig, giving it a handler that, in turn, uses a structlog formatter
2. Using Celery's signals to instruct it to use said logging configuration on bootup

With those two accomplished, `wrap_logger()` works perfectly with this hopefully minimal-enough example:

```python3
# my_project/log_config.py
# module used both by Flask app and Celery worker
import structlog
from logging.config import dictConfig

dict_config = {
    "version": 1,
    "disable_existing_loggers": False,
    "formatters": {
        "dev_console": {
            "()": structlog.stdlib.ProcessorFormatter,
            "processor": structlog.dev.ConsoleRenderer(),
        },
    },
    "handlers": {
        "console": {
            "class": "logging.StreamHandler",
            "formatter": "dev_console",
        },
    },
    "loggers": {
        "my_project": {
            "level": "INFO",
        },
        # Key #1: This is what I needed since my root is WARN-level. Ultimately,
        # explicitly calling out `celery` or `celery.task` in your logging
        # config is critical to controlling how it responds to your structlog
        # setup.
        "celery.task": {
            "level": "INFO",
        },
    },
    "root": {
        "level": "WARNING",
        "handlers": ["console"],
    },
}

def configure_logging() -> None:
    dictConfig(dict_config)
    structlog.configure(
        processors=[
            # ...
            structlog.stdlib.ProcessorFormatter.wrap_for_formatter,
        ],
        wrapper_class=structlog.stdlib.BoundLogger,
        logger_factory=structlog.stdlib.LoggerFactory(),
        # ...
    )
```

```python3
# my_project/tasks.py

import structlog
import celery
from celery import signals
from my_project.log_config import configure_logging

celery = Celery(...)

# Key #2: Celery *needs* you to use one of its signals to configure logging. I
# chose this one, which overrides any setup they would do. A wiser and more
# patient person would consider `after_setup_task_logger` to augment.
@signals.setup_logging.connect
def on_setup_logging(**kwargs):
    configure_logging()

# all tasks are defined from here on out. Would normally live in another module
from celery.utils.log import get_task_logger
logger = structlog.wrap_logger(get_task_logger(__name__))

@celery.task()
def async_add(x, y):
    res = x + y
    logger.info("Performing addition", x=x, y=y, res=res)
    return x + y
```

**TLDR:** explicitly configuring for `celery.task` logger in my stdlib-using-structlog configuration, and then making sure to call that configuration setup during the celery `setup_logging` signal, is what allowed me to use `structlog.wrap_logger(get_task_logger(__name__))` so I can develop happily ever after.

**Comment by f3ndot** [comment]
For those who are curious, I also found I am able to take advantage of more Celery signals to bind the task id and name in a thread local, not unlike one can do in a Flask request.

Expanding the above example:

```python3
# Be sure to add `structlog.threadlocal.merge_threadlocal` to the top of your structlog.configure's processors chain

@signals.task_prerun.connect
def on_task_prerun(sender, task_id, task, args, kwargs, **_kwargs):
    clear_threadlocal() # a just-in-case
    bind_threadlocal(task_id=task_id, task_name=task.name)


@signals.task_postrun.connect
def on_task_postrun(sender, task_id, task, args, kwargs, retval, state, **_kwargs):
    clear_threadlocal()
```

You can now use your logger as before;

```python3
logger = structlog.wrap_logger(get_task_logger(__name__))

@celery.task()
def async_add(x, y):
    res = x + y
    logger.info("Performing addition", x=x, y=y, res=res)
    return x + y
```

with the output being (assuming timestamp, logger name, and other pretty processors configured):

```
2021-12-10 13:02.41 [info     ] Performing addition            [my_project.tasks] res=12 task_id=cb1975e9-d808-4e4b-abd6-5f1bd3314ae6 task_name=my_project.tasks.async_add x=3 y=4```

**Comment by hynek** [comment]
@f3ndot could I convince you to contribute your wisdom to <https://github.com/hynek/structlog/blob/main/docs/recipes.md>?

Additionally it would be great, if you could help out here: https://stackoverflow.com/questions/74004852/race-condition-and-interleaving-when-struclog-logger-is-used-within-celery-tasks

I've tried but it's getting too specific … TIA!
